linux/daemon/OSPDaemon.c

```diff
@@ -0,0 +1,597 @@
+
+	nstart = start;
+	for (i = 0; i < sd->output_count; i++) {
+		nstart++;
+		pfd[i+start].fd = sd->output[i].fd;
+		pfd[i+start].events = 0;
+		OSPDaemon_queue_init(&sd->output[i].q);
+
+		if (sd->output[i].noprocess) continue;
+
+		sd->output[i].ResultDesc.pResultReadyCallback = OSPDaemon_callback;
+		if (sd->output[i].type != -1) {
+			oret = OSP_SubscribeSensorResult(&sd->output[i].ResultDesc, &sd->output[i].handle);
+			if (oret != OSP_STATUS_OK) {
+				DBG(DEBUG_INIT, "Sensor subscribe failed\n");
+			}
+		}
+	}
+	return nstart;
+}
+
+/* 
+ * start - index to start populating the pfd array.
+ *
+ * Returns start + number of entries used.
+ */
+static int OSPDaemon_init_inbound(struct pollfd *pfd, int nfd, int start)
+{
+	int i;
+	OSP_STATUS_t oret;
+	int nstart = start;
+
+	if (sd->CalFile)
+		ReadCalData(sd->CalFile);
+
+	for (i = 0; i < sd->sensor_count; i++) {
+		nstart++;
+		if (sd->sensor[i].fd < 0)
+			continue;
+
+		pfd[i+start].fd = sd->sensor[i].fd;
+		pfd[i+start].events = POLLIN;
+		OSPDaemon_queue_init(&sd->sensor[i].q);
+		DBG(DEBUG_INIT, "Input sensor noprocess = %i\n",
+				sd->sensor[i].noprocess);
+		if (sd->sensor[i].noprocess) continue;
+		sd->sensor[i].sensor.pOptionalWriteDataCallback =
+			OSPDaemon_write_cal_cb;
+
+		oret = OSP_RegisterInputSensor(&sd->sensor[i].sensor, &sd->sensor[i].handle);
+		if (oret != OSP_STATUS_OK) {
+			DBG(DEBUG_INIT, "Sensor register failed (%i)\n", oret);
+		}
+	}
+	return nstart;
+}
+
+static void OSPDaemon(char *confname)
+{
+	int i, r = 0;
+	int dirty;
+
+	if ((sd = OSPDaemon_config(confname)) == NULL) {
+		fprintf(stderr, "Bad/invalid config.\n");
+		return;
+	}
+
+	OSP_Initialize(&SystemDesc);
+
+	for (i = 0; i < nfd; i++) {
+		pfd[i].fd = 0;
+		pfd[i].events = 0;
+	}
+
+	OSPDaemon_driver_setup_in(sd->sensor, sd->sensor_count);
+	r = OSPDaemon_init_inbound(pfd, nfd, r);
+	OSPDaemon_driver_setup_out(sd->output, sd->output_count);
+	nfd = OSPDaemon_init_outbound(pfd, nfd, r);
+	sd->powerfd = OSPDaemon_power_init(sd);	
+	if (sd->powerfd > 0) {
+		pfd[nfd].fd = sd->powerfd;
+		pfd[nfd].events |= POLLIN;
+		nfd++;
+	}
+
+	for (i = 0; i < nfd; i++) {
+		DBG(DEBUG_INIT, "%i: fd %i events = %i\n", i, pfd[i].fd, pfd[i].events);
+	}
+
+	dirty = 0;
+	while (!mainstatus) {
+		if (poll(pfd, nfd, -1) <= 0)
+			continue;
+
+		for (i = 0; i < nfd; i++) {
+			if (pfd[i].revents & POLLIN) {
+				DBG(DEBUG_LOOP, "IN event\n");
+				if (pfd[i].fd == sd->powerfd) {
+					OSPDaemon_power_process(sd,
+								sd->powerfd);
+				} else {
+					dirty = OSPDaemon_process_inbound(
+								&pfd[i], pfd,
+								nfd, dirty);
+				}
+			}
+			if (pfd[i].revents & POLLOUT) {
+				DBG(DEBUG_LOOP, "%s:%i OUT event\n", __func__, __LINE__);
+				OSPDaemon_process_outbound(&pfd[i], nfd);
+			}
+		}
+
+		OSPDaemon_flush_out(sd);
+
+		if (dirty) {
+			DBG(DEBUG_LOOP, "%s:%i PROCESS\n", __func__, __LINE__);
+			if (OSP_DoForegroundProcessing() == OSP_STATUS_IDLE)
+				dirty = 0;
+			OSP_DoBackgroundProcessing();
```

Comment:
OSP_DoBackgroundProcessing, OSP_DoForegroundProcessing - These are blocking calls, and take a constant amount of time. Is it possible to still maintain the sampling period requested by Android with this additional processing?